test/parallel_api/iterator/permutation_iterator.pass.cpp

```diff
@@ -59,7 +60,7 @@ struct is_able_to_to_modify_src_data_in_test : ::std::true_type { };
 #if TEST_DPCPP_BACKEND_PRESENT
 template <typename ExecutionPolicy>
 struct is_able_to_to_modify_src_data_in_test<ExecutionPolicy, perm_it_index_tags::counting>
```

Comment:
@danhoeflinger could you please explain why you have changed the source implementations of  `struct is_able_to_to_modify_src_data_in_test` in test and exactly in which cases our previous implementation were incorrect and why?
Thanks.

Reply:
Sure.  Let me first state my understanding of `is_able_to_to_modify_src_data_in_test` generally.  (which I now notice has a typo `..._to_to_...`)

`is_able_to_to_modify_src_data_in_test` is a struct which determines if we are able to modify base iterator data within a kernel which has a specific execution policy and a specific index iterator type (and therefore if we should run the test with this configuration).  I believe we assume that the base iterator data is modify-able using the selected execution policy by our test setup.  Therefore, what is important is if we can safely dereference the index iterator type in a kernel associated with the execution policy specified.  

Now on to my changes:
`is_able_to_to_modify_src_data_in_test<ExecutionPolicy, perm_it_index_tags::counting>` -> `std::true_type`.
A counting iterator is passed directly, and should be usable for both host and device kernels.  Previously we were skipping all device execution policy tests which used counting iterator indices for permutation iterator tests, with my changes they are no longer skipped.

`is_able_to_to_modify_src_data_in_test<ExecutionPolicy, perm_it_index_tags::usm_shared>` -> `std::true_type`.
usm_shared memory should be accessible both on the host and device.  Previously we were skipping all device execution policy tests for types using usm_shared index iterators.  With my changes they are no longer skipped.

`is_able_to_to_modify_src_data_in_test<ExecutionPolicy, perm_it_index_tags::usm_device>` -> `oneapi::dpl::__internal::__is_hetero_execution_policy<::std::decay_t<ExecutionPolicy>>`.
usm_device memory should be accessible only on the device.  Previously we were skipping all device execution policy tests for types using usm_device index iterators, but actually permitting it for host side policies.  With my changes, we  test with usm_device index iterators, but only on the device.

Reply:
@MikeDvorskiy could you please take a look to this change?

Reply:
> As far as I remember we are unable to modify source data (it test this mean - unable to call algorithms with in-place data modification) for hetero policy with permutation iterator. And previous implementation were the same for all tags, it's true.

I'm not sure why this would be the case, assuming the source data is a valid supported type for permutation iterator.